clippy_lints/src/needless_move.rs

```diff
@@ -0,0 +1,222 @@
+            // async fn ...() {} convert the body to an `async move {}` block,
+            // with a DUMMY_SP for the move_kw
+            return;
+        }
+
+        // Collect moved & borrowed variables from the closure, which the closure *actually* needs.
+        let ctx = {
+            let mut ctx = MovedVariablesCtxt::default();
+            let body = cx.tcx.hir().body(closure.body);
+            let infcx = cx.tcx.infer_ctxt().build();
+            euv::ExprUseVisitor::new(&mut ctx, &infcx, closure.def_id, cx.param_env, cx.typeck_results())
+                .consume_body(body);
+            ctx
+        };
+
+        let mut lint_result = LintResult::NothingCaptured;
+
+        for captured_place in cx.typeck_results().closure_min_captures_flattened(closure.def_id) {
+            let place = &captured_place.place;
+            if let Some(ck_expr_id) = captured_place.info.capture_kind_expr_id {
+                let required_ck = ctx.get_required_kind(place, ck_expr_id);
+                match required_ck {
+                    UpvarCapture::ByRef(_) => {
+                        // no matter what the old `lint_result` is, we keep the move.
+                        lint_result = LintResult::NeedMove;
+                    },
+                    UpvarCapture::ByValue => {
+                        lint_result = match lint_result {
+                            LintResult::NothingCaptured | LintResult::Consumed => LintResult::Consumed,
+                            LintResult::NeedMove => LintResult::NeedMove,
+                        }
+                    },
+                }
+            }
+        }
+
+        let lint = |note_msg: &'static str| {
+            span_lint_and_then(
+                cx,
+                NEEDLESS_MOVE,
+                expr.span,
+                "you seem to use `move`, but the `move` is unnecessary",
+                |diag| {
+                    diag.suggest_remove_item(cx, move_kw, "remove the `move`", Applicability::MachineApplicable);
+                    diag.note(note_msg);
+                },
+            );
+        };
+
+        match lint_result {
+            LintResult::NothingCaptured => {
+                lint("there were no captured variables, so the `move` is unnecessary");
+            },
+            LintResult::Consumed => {
+                lint("there were consumed variables, but no borrowed variables, so the `move` is unnecessary");
+            },
+            LintResult::NeedMove => {
+                // there was a value which would be borrowed if it weren't for the move keyword,
+                // so we should keep it, as removing it would change semantics.
+            },
+        }
+    }
+}
+
+impl<'tcx> LateLintPass<'tcx> for NeedlessMove {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
+        if expr.span.from_expansion() {
+            return;
+        }
+
+        let ExprKind::Closure(closure) = &expr.kind else {
```

Comment:
This could be a let-chain instead. However since this is only ever called once, maybe you can inline it?